File: agent/database/engine.py
"""
Database engine and connection management.

Single SQLite database for:
- Conversation metadata (our table)
- LangGraph checkpoints (AsyncSqliteSaver tables)
"""

import os
import logging
from pathlib import Path
from typing import Optional

from sqlalchemy.ext.asyncio import (
    create_async_engine,
    AsyncSession,
    async_sessionmaker,
    AsyncEngine,
)

from .models import Base

logger = logging.getLogger("movesia.database")

# Global instances (initialized during startup)
_engine: Optional[AsyncEngine] = None
_session_factory: Optional[async_sessionmaker[AsyncSession]] = None
_checkpoint_saver = None
_checkpoint_conn = None  # aiosqlite connection for checkpoint saver


def get_database_path() -> str:
    """
    Get the path to the SQLite database file.

    Uses DATABASE_PATH env var if set, otherwise defaults to
    'data/movesia.db' in the agent directory.
    """
    if db_path := os.getenv("DATABASE_PATH"):
        return db_path

    # Default: agent/data/movesia.db
    agent_dir = Path(__file__).parent.parent
    data_dir = agent_dir / "data"
    data_dir.mkdir(exist_ok=True)
    return str(data_dir / "movesia.db")


async def init_database() -> async_sessionmaker[AsyncSession]:
    """
    Initialize the database engine and create tables.

    Call this during FastAPI startup.
    Returns the session factory for dependency injection.
    """
    global _engine, _session_factory, _checkpoint_saver

    db_path = get_database_path()
    logger.info(f"Initializing database at: {db_path}")

    # Create async engine
    _engine = create_async_engine(
        f"sqlite+aiosqlite:///{db_path}",
        echo=False,
        connect_args={"check_same_thread": False},
    )

    # Create session factory
    _session_factory = async_sessionmaker(
        _engine,
        class_=AsyncSession,
        expire_on_commit=False,
    )

    # Create our tables
    async with _engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    logger.info("Database tables created/verified")

    # Initialize LangGraph checkpoint saver (same database)
    global _checkpoint_conn
    try:
        from langgraph.checkpoint.sqlite.aio import AsyncSqliteSaver
        import aiosqlite

        # AsyncSqliteSaver.from_conn_string returns a context manager
        # We need to create the connection and saver manually
        _checkpoint_conn = await aiosqlite.connect(db_path)
        _checkpoint_saver = AsyncSqliteSaver(_checkpoint_conn)
        await _checkpoint_saver.setup()  # Create checkpoint tables
        logger.info("LangGraph AsyncSqliteSaver initialized (same DB)")
    except ImportError as e:
        logger.warning(
            f"langgraph-checkpoint-sqlite or aiosqlite not installed: {e}. "
            "Install with: pip install langgraph-checkpoint-sqlite aiosqlite"
        )
        from langgraph.checkpoint.memory import MemorySaver
        _checkpoint_saver = MemorySaver()
        logger.info("Falling back to MemorySaver (no persistence)")

    return _session_factory


async def close_database() -> None:
    """Close database connections gracefully."""
    global _engine, _session_factory, _checkpoint_saver, _checkpoint_conn

    if _checkpoint_conn:
        await _checkpoint_conn.close()
        logger.info("Checkpoint connection closed")

    if _engine:
        await _engine.dispose()
        logger.info("Database engine disposed")

    _engine = None
    _session_factory = None
    _checkpoint_saver = None
    _checkpoint_conn = None


def get_session_factory() -> async_sessionmaker[AsyncSession]:
    """Get the session factory (must call init_database first)."""
    if _session_factory is None:
        raise RuntimeError("Database not initialized. Call init_database() first.")
    return _session_factory


def get_database() -> AsyncEngine:
    """Get the database engine (must call init_database first)."""
    if _engine is None:
        raise RuntimeError("Database not initialized. Call init_database() first.")
    return _engine


def get_checkpoint_saver():
    """Get the LangGraph checkpoint saver."""
    if _checkpoint_saver is None:
        raise RuntimeError("Database not initialized. Call init_database() first.")
    return _checkpoint_saver


class DatabaseSession:
    """
    Async context manager for database sessions.

    Usage:
        async with DatabaseSession() as session:
            result = await session.execute(query)
    """

    def __init__(self):
        self._session: Optional[AsyncSession] = None

    async def __aenter__(self) -> AsyncSession:
        factory = get_session_factory()
        self._session = factory()
        return self._session

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._session:
            if exc_type is not None:
                await self._session.rollback()
            await self._session.close()

File: src/agent/database/engine.ts
/**
 * Database engine and connection management using sql.js.
 *
 * Uses sql.js (WASM SQLite) instead of better-sqlite3 to avoid
 * native module compilation issues in VS Code extensions.
 *
 * Single SQLite database for:
 * - Conversation metadata (our table)
 * - LangGraph checkpoints (SqlJsCheckpointer tables)
 *
 * This module is designed to run inside the VS Code extension process.
 * Database path is set dynamically from VS Code's globalStorageUri.
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'fs';
import { dirname, join } from 'path';
import initSqlJs, { type Database as SqlJsDatabase } from 'sql.js';
import { createLogger } from '../UnityConnection/config';
import { CONVERSATIONS_SCHEMA } from './models';
import { SqlJsCheckpointer } from './SqlJsCheckpointer';

const logger = createLogger('movesia.database');

// Global instances (initialized during startup)
let _db: SqlJsDatabase | null = null;
let _checkpointSaver: SqlJsCheckpointer | null = null;
let _sqlJs: Awaited<ReturnType<typeof initSqlJs>> | null = null;

// Database path - set from VS Code extension context
let _storagePath: string | null = null;
let _dbPath: string | null = null;

/**
 * Set the storage path from VS Code's globalStorageUri.
 * Call this from the extension before initializing the database.
 */
export function setStoragePath(path: string): void {
    _storagePath = path;
}

/**
 * Get the path to the SQLite database file.
 *
 * Priority:
 * 1. DATABASE_PATH env var (for testing)
 * 2. VS Code storage path (set via setStoragePath)
 * 3. Fallback to temp directory
 */
export function getDatabasePath(): string {
    const envPath = process.env.DATABASE_PATH;
    if (envPath) {
        return envPath;
    }

    if (_storagePath) {
        // Ensure storage directory exists
        if (!existsSync(_storagePath)) {
            mkdirSync(_storagePath, { recursive: true });
        }
        return join(_storagePath, 'movesia.db');
    }

    // Fallback: use temp directory
    const tempDir = process.env.TEMP || process.env.TMP || '/tmp';
    const dataDir = join(tempDir, 'movesia-data');

    if (!existsSync(dataDir)) {
        mkdirSync(dataDir, { recursive: true });
    }

    logger.warn('No storage path set, using temp directory: ' + dataDir);
    return join(dataDir, 'movesia.db');
}

/**
 * Save the database to disk.
 */
function persistDatabase(): void {
    if (_db && _dbPath) {
        const data = _db.export();
        const buffer = Buffer.from(data);
        writeFileSync(_dbPath, buffer);
    }
}

/**
 * Initialize the database and create tables.
 *
 * Call this during server startup.
 */
export async function initDatabase(): Promise<SqlJsDatabase> {
    _dbPath = getDatabasePath();
    logger.info(`Initializing database at: ${_dbPath}`);

    // Ensure directory exists
    const dbDir = dirname(_dbPath);
    if (!existsSync(dbDir)) {
        mkdirSync(dbDir, { recursive: true });
    }

    // Initialize sql.js
    _sqlJs = await initSqlJs();

    // Create or load database
    if (existsSync(_dbPath)) {
        const buffer = readFileSync(_dbPath);
        _db = new _sqlJs.Database(buffer);
        logger.info('Loaded existing database');
    } else {
        _db = new _sqlJs.Database();
        logger.info('Created new database');
    }

    // Create our tables
    _db.run(CONVERSATIONS_SCHEMA);
    logger.info('Database tables created/verified');

    // Persist initial state
    persistDatabase();

    // Initialize LangGraph checkpoint saver using sql.js
    _checkpointSaver = await SqlJsCheckpointer.create(_dbPath);
    logger.info('SqlJsCheckpointer initialized (persistent)');

    return _db;
}

/**
 * Close database connections gracefully.
 */
export async function closeDatabase(): Promise<void> {
    if (_checkpointSaver) {
        _checkpointSaver.close();
        logger.info('Checkpointer closed');
    }

    if (_db) {
        persistDatabase();
        _db.close();
        logger.info('Database connection closed');
    }

    _db = null;
    _checkpointSaver = null;
    _sqlJs = null;
}

/**
 * Get the database instance (must call initDatabase first).
 */
export function getDatabase(): SqlJsDatabase {
    if (_db === null) {
        throw new Error('Database not initialized. Call initDatabase() first.');
    }
    return _db;
}

/**
 * Get the LangGraph checkpoint saver.
 */
export function getCheckpointSaver(): SqlJsCheckpointer {
    if (_checkpointSaver === null) {
        throw new Error('Database not initialized. Call initDatabase() first.');
    }
    return _checkpointSaver;
}

/**
 * Check if database is initialized.
 */
export function isDatabaseInitialized(): boolean {
    return _db !== null;
}

/**
 * Save database changes to disk.
 * Call this after write operations.
 */
export function saveDatabase(): void {
    persistDatabase();
}
